### PR TITLE
hw-02: rest with crud

### PR DESCRIPTION
**Pull request by aliskar**
### TASK 2.1
Write a simple **REST** service with **CRUD** operations for **User** entity.
- To create REST service, use ExpressJS (https://expressjs.com/).
The `User` should have the following properties (you can use **UUID** as a user identifier (**id**)):
```
type User = {
  id: string;
  login: string;
  password: string;
  age: number;
  isDeleted: boolean;
};
```
- Service should have the following **CRUD** operations for User:
  - get user by id;
  - create and update user;
  - get auto-suggest list from `limit` users, sorted by login property and filtered by loginSubstring in the login property:
 `getAutoSuggestUsers(loginSubstring, limit)`
  - emove user (soft delete – user gets marked with `isDeleted` flag, but not removed from the collection).

- Store user’s collection in the service memory (while the service is running).
To test the service **CRUD** methods, you can use **Postman** (https://www.getpostman.com/).

### TASK 2.2
Add server-side validation for create/update operations of User entity:
- all fields are required;
- login validation is required;
- password must contain letters and numbers;
- user’s age must be between 4 and 130.

In case of any property does not meet the validation requirements or the field is absent, return **400** (Bad Request) and detailed error message.
For requests validation use special packages like **joi** (https://github.com/hapijs/joi, https://www.npmjs.com/package/express-joi-validation).

### CODE REVIEW

File: src/hw_02/README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+POST:`localhost:3000/users/:id`
+* will return all existing users with new one into `json`.
+* next fields are required and have to be valid :
+    * login;
+    * password must contain minimum eight characters, at least one letter and one number;
+    * age must be between 4 and 130;
+
+### Get suggested user
+
+GET:`localhost:3000/suggestUsers?partialLogin=d&limit=5`
```
Comment (Metre-Ray) [inline src/hw_02/README.md:24]: please, fix description, the correct url is `localhost:3000/users/suggest?limit=8&partialLogin=e`

File: src/hw_02/routes/routes.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  if (requestedUser) {
+    res.json(requestedUser);
+  } else {
+    res.status(404).send('The requested user is not found');
+  }
+});
+
+router.post('/', validationMiddleware(UserSchema), (req, res) => {
+  users.push({ ...req.body, id: uuid(), isDeleted: false });
+  res.json(getAllExistingUsers(users));
```
Comment (Metre-Ray) [inline src/hw_02/routes/routes.ts:41]: just a note, after creation usually the created entity is sent back, not the whole list of entities
also, the suitable status here is 201

File: src/hw_02/README.md (hunk 1 above)
Comment (aliskar) [reply src/hw_02/README.md]: done

File: src/hw_02/routes/routes.ts (hunk 2 above)
Comment (aliskar) [reply src/hw_02/routes/routes.ts]: done

**Pull request merged**